### ISSUE TITLE
[Bug]: Python SDK: Dependency installation hangs sometimes

### ISSUE THREAD

**Issue by phoerious** [apache/beam#24357]
### What happened?

Disclaimer: This bug reports refers to the current master branch, which contains major changes to the container boot.go, but it's very likely that it applies to previous versions as well, since the parts which most likely trigger the bug were not changed.

The Python SDK boot.go binary installs dependencies from a setup.py automatically here: https://github.com/apache/beam/blob/master/sdks/python/container/boot.go#L368

Most of the time, it works fine, but sometimes `pip` process seems to hang for no reason. The problem can be resolved by attaching to the container running the SDK and reading STDIN of the `boot` process:

```kubectl exec -ti pod/beam-flinker-cluster-taskmanager -c beam-worker-pool -- bash
ps aux | grep boot
cat /proc/<WORKER-PID>/fd/1
exit
```

I suppose this issue occurs due to a stdout buffer fillup. The boot binary uses the following utility for running `pip`, which redirects the processe's STDOUT to `os.stdout`. A solution would perhaps be to either not redirect stdout or to write it directly to `/dev/null`.

https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/util/execx/exec.go#L26

### Issue Priority

Priority: 2

### Issue Component

Component: sdk-py-harness

**Comment by tvalentyn** [comment]
I frequently debug worker startup issues on Dataflow (which also uses containers), and have not seen dependency installation to just hang for no reason. I don't think redirecting to /dev/null is an option as the installation logs are useful. Are you able to repro the issue consistently? Are there any logs that demonstrate the stuckness to rule out that it is not caused by dependency issues?

**Comment by phoerious** [comment]
No, it does not happen consistently. it is probably dependent on the host's utilisation or IO saturation. I also don't think it's Flink-related. There is nothing special in the logs. It usually hangs after pip's last installation log dump, just before the final "Successfully installed pkg1 pkg2 ..." line. Reading pip's or the worker's stdout via `cat` will let it continue like nothing happened.

Redirecting everything `/dev/null` may be a bit harsh, but redirecting pip's output to the worker's stdout is definitely useless. What is shown as the container's log output is the worker pool output, which displays only the pool's own stdout and log.Printf() messages as well as log.Printf() messages from the workers (apparently). It does not include normal stdout of child processes, including the workers. Therefore no pip output gets logged anywhere, really. In order to read or log it, you have to attach to the container and read from `/proc/PID/fd/1` directly.

**Comment by tvalentyn** [comment]
> In order to read or log it, you have to attach to the container

Dataflow might be doing that, perhaps that is why we never observe such behavior.

> I'll keep testing it

Thanks. If you can confirm that this behavior is unrelated to your change, that would be helpful. If you suspect logging is an issue, you can intentionally flood the logs a little to repro.

**Comment by phoerious** [comment]
I did not change how pip is being called, so I highly doubt that it is related to my changes.

> Dataflow might be doing that, perhaps that is why we never observe such behavior.

But does it also explicitly flush the stdout of all processes? Kubelet also reads stdout of the container, but pip is not part of the container stdout.

**Comment by tvalentyn** [comment]
we do get pip's output in Dataflow startup logs.

**Comment by tvalentyn** [comment]
> I did not change how pip is being called

at minimum I think you changed invocation to use: `python -m pip` instead of `pip`.

**Comment by phoerious** [comment]
Which is the exact equivalent of what was there before, only safer.

**Comment by tvalentyn** [comment]
just checking if you might have any new information here. thanks.

**Comment by phoerious** [comment]
My guess is that this is just I/O wait because redirecting stdio of the process has a limited buffer size. Why this usually seems to affect only the very last line of output, I don't know, but that may be coincidence, since the pip output length is usually rather predictable.

**Comment by phoerious** [comment]
@tvalentyn I think I may have found why this is happening. The following line connects the container executable's STDOUT and then simply starts the process without interacting with it any further. This will fill up the internal I/O buffers and hang forever unless the output is consumed somehow, e.g. by calling `communicate()`. https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/worker/worker_pool_main.py#L167

This behaviour is documented: https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait

This is a general problem with pipes, so it's not specific to Python. These downstream lines in Go are just as problematic and may already cause deadlocks if the upstream Python call doesn't: https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/util/execx/exec.go#L34-L36

See also https://devblogs.microsoft.com/oldnewthing/20110707-00/?p=10223

**Comment by tvalentyn** [comment]
Interesting.

>  These downstream lines in Go are just as problematic

I don't think we observed deadlocks in Dataflow environments but cc'ing: @lostluck our Go expert who may have suggestions and @jrmccluskey who is looking at improving logging in boot loaders.

**Comment by lostluck** [comment]
From pkg.go.dev/os

`Stdin, Stdout, and Stderr are open Files pointing to the standard input, standard output, and standard error file descriptors.`

So it's technically whatever the OS should be doing by default.

I've suggested to Jack that we should swap those with our own buffered IO so we can parse & emit them to the logging service.

But i wouldn't suggest we try to get that change in the day before the 2.50 cut.

**Comment by robertwb** [comment]
Regarding the lines at worker_pool_main.py#L16, these shouldn't be used in the dataflow service (i.e. use_subprocess, which is primarily for testing, should be false).

(I'd also be surprised if os.Stdout, etc. limited buffers.) 

If we capture/redirect this somewhere, we should be very careful, as otherwise that can make debugging very difficult.
